classes/manager.php

```diff
@@ -257,12 +261,101 @@ public static function set_pass_state() {
             $event = \tool_mfa\event\user_passed_mfa::user_passed_mfa_event($USER);
             $event->trigger();
 
+            // Unset session vars during mfa auth.
+            unset($SESSION->mfa_redir_referer);
+            unset($SESSION->mfa_redir_count);
+
             // Fire post pass state factor actions.
             $factors = \tool_mfa\plugininfo\factor::get_active_user_factor_types();
             foreach ($factors as $factor) {
                 $factor->post_pass_state();
             }
         }
     }
-}
 
+    /**
+     * Checks whether the user should be redirected from the provided url.
+     *
+     * @return int
+     */
+    public static function should_require_mfa($url, $preventredirect) {
+        global $CFG, $USER, $SESSION;
+        // Remove all params before comparison.
+        $url->remove_all_params();
+
+        // Honor prevent_redirect.
+        if ($preventredirect) {
+            return 0;
+        }
+
+        // WS/AJAX check.
+        if (WS_SERVER || AJAX_SCRIPT) {
+            return self::REDIRECT_EXCEPTION;
+        }
+
+        // User/edit.php.
+        $useredit = new \moodle_url('/user/edit.php');
```

Comment:
I think this needs to be more like if (user_not_full_setup() then NO_REDIRECT

Test this in profile editing when some ajax calls